File: src/main/java/io/bespin/demo/mapreduce/pagerank/RangePartitioner.java
package io.bespin.demo.mapreduce.pagerank;

import org.apache.hadoop.conf.Configurable;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Writable;
import org.apache.hadoop.mapreduce.Partitioner;

/**
 * Ranger partitioner. In the context of graph algorithms, ensures that consecutive node ids are
 * blocked together.
 *
 * @author Jimmy Lin
 * @author Michael Schatz
 */
public class RangePartitioner extends Partitioner<IntWritable, Writable> implements Configurable {
  private int nodeCnt = 0;
  private Configuration conf;

  public RangePartitioner() {}

  @Override
  public int getPartition(IntWritable key, Writable value, int numReduceTasks) {
    return (int) (((float) key.get() / (float) nodeCnt) * numReduceTasks) % numReduceTasks;
  }

  @Override
  public Configuration getConf() {
    return conf;
  }

  @Override
  public void setConf(Configuration conf) {
    this.conf = conf;
    configure();
  }

  private void configure() {
    nodeCnt = conf.getInt("NodeCount", 0);
  }
}

File: src/main/scala/io/bespin/scala/mapreduce/pagerank/RangePartitioner.scala
package io.bespin.scala.mapreduce.pagerank

import org.apache.hadoop.conf.{Configurable, Configuration}
import org.apache.hadoop.io.{IntWritable, Writable}
import org.apache.hadoop.mapreduce.Partitioner


class RangePartitioner[T <: Writable] extends Partitioner[IntWritable, T] with Configurable {
  private var nodeCnt: Int = 0
  private var conf: Configuration = null

  def getPartition(key: IntWritable, value: T, numReduceTasks: Int): Int = {
    ((key.get.toFloat / nodeCnt.toFloat) * numReduceTasks).toInt % numReduceTasks
  }

  def getConf: Configuration = {
    conf
  }

  def setConf(conf: Configuration): Unit = {
    this.conf = conf
    configure()
  }

  private def configure(): Unit = {
    nodeCnt = conf.getInt("NodeCount", 0)
  }
}
